### ISSUE TITLE
Issue installing EasyBuild with the bootstrap method

### ISSUE THREAD

**Issue by sylmarien** [hpcugent/easybuild-framework#1089]
Hi,

Since lmod has just been packaged in a compatible version for Debian ([lmod package on Debian unstable](https://packages.debian.org/sid/lmod)) I'm trying to use it instead of environment-modules on Debian 7.6.

But the installation fails during STAGE 2 (during the sanity check more precisely) with this error:
```
== sanity checking...
== FAILED: Installation ended unsuccessfully (build directory: /tmp/tmpwkRfru/EasyBuild/1.15.2/dummy-dummy): build failed (first 300 chars): EasyBuild crashed with an error (at easybuild/framework/easyblock.py:1612 in sanity_check_step): Sanity check failed: EasyBuild failed to install, cmd 'python -c "import easybuild"' (stdin: None) output: Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module
== Results of the build can be found in the log file /tmp/easybuild-Ih3Qma/easybuild-EasyBuild-1.15.2-20141106.091730.VcgNZ.log
Traceback (most recent call last):
  File "bootstrap_eb.py", line 584, in <module>
    main()
  File "bootstrap_eb.py", line 411, in main
    stage2(tmpdir, versions, install_path)
  File "bootstrap_eb.py", line 360, in stage2
    easybuild_main()
  File "/tmp/tmpwkRfru/eb_stage1/lib/python2.7/site-packages/easybuild_framework-1.15.2-py2.7.egg/easybuild/main.py", line 441, in main
    ecs_with_res = build_and_install_software(ordered_ecs, init_session_state, exit_on_failure=exit_on_failure)
  File "/tmp/tmpwkRfru/eb_stage1/lib/python2.7/site-packages/easybuild_framework-1.15.2-py2.7.egg/easybuild/main.py", line 112, in build_and_install_software
    _log.error(test_msg)
  File "/tmp/tmpwkRfru/eb_stage1/lib/python2.7/site-packages/easybuild_framework-1.15.2-py2.7.egg/easybuild/tools/build_log.py", line 105, in error
    raise EasyBuildError(newMsg)
easybuild.tools.build_log.EasyBuildError: 'EasyBuild crashed with an error (at easybuild/main.py:112 in build_and_install_software): Build of /tmp/tmpwkRfru/EasyBuild-1.15.2.eb failed (err: \'build failed (first 300 chars): EasyBuild crashed with an error (at easybuild/framework/easyblock.py:1612 in sanity_check_step): Sanity check failed: EasyBuild failed to install, cmd \\\'python -c "import easybuild"\\\' (stdin: None) output: Traceback (most recent call last):\\n  File "<string>", line 1, in <module>\\nImportError: No module\')'
Exception TypeError: "'NoneType' object is not callable" in <function _removeHandlerRef at 0x39f8b90> ignored
```

(Here is the full log for more informations: [error log on gist](https://gist.github.com/sylmarien/454aa3fa194874332da4)

To precise a bit the context:  
The command I use is:  
`$ python bootstrap_eb.py /opt/apps/EasyBuild`
The output of the command asked in the installation page are:  
```
$ python -V
Python 2.7.3
```
```
$ type -f module
bash: type: module: not found
```
(But `type module` output is correct:  
```
$ type module
module is a function
module () 
{ 
    eval $($LMOD_CMD bash "$@");
    [ $? = 0 ] && eval $(${LMOD_SETTARG_CMD:-:} -s sh)
}
```
)
```
$ module --version

Modules based on Lua: Version 5.8  2014-11-04 12:46
    by Robert McLay mclay@tacc.utexas.edu
```
```
$ module av EasyBuild

Use "module spider" to find all possible modules.
Use "module keyword key1 key2 ..." to search for all possible modules matching any of the "keys".
```  
`$ which -a eb` has no output  
```
$ eb --version
bash: eb: command not found
```

So I'm not sure if this is a bug or something I did wrong, moreover, if this is a bug, I don't know if it is specific to the Debian 7.6 or to lmod 5.8, since when I do the exact same procedure on Centos 7.0 with lmod 5.7.5 everything works fine.

**Comment by boegel** [comment]
@sylmarien: sorry for never getting back to this

Lmod does require Tcl to be installed, since it also supports Tcl module files (i.e. the ones produced by EasyBuild). Lmod will translate the Tcl to Lua internally, and then continue with parsing the Lua module file, using a tool named `tcl2lua`. It needs `tclsh` to be around, at that point.

Thanks for pointing this out, I'll try and look into updating the EasyBuild documentation accordingly.

**Comment by boegel** [comment]
@rtmclay: shouldn't Lmod make sure that `tclsh` is around somehow at configure time?

**Comment by rtmclay** [comment]
But you do not have to have tclsh to make Lmod work.  You can do a pure lua solution, you just can't deal with any TCL modulefiles.

**Comment by boegel** [comment]
@rtmclay: if `--wtih-tcl` is enabled (which is the default), then the configure should be checking for `tclsh`, no? otherwise the Tcl support in Lmod can never work?

**Comment by rtmclay** [comment]
Lmod has checked for tclsh in the path.  I have modified the configure step to die if it can't find tclsh and --with-tcl is yes

**Comment by boegel** [comment]
Makes sense, thanks @rtmclay!

@sylmarien: so, future version of Lmod should complain if `tclsh` is missing, I'll leave this open until I've updated the EB docs to mention that Tcl is also required with Lmod in case you'll be using Tcl module files (only option with EB, for now)

**Comment by sylmarien** [comment]
@boegel Ok, thanks for the explanations and thank you for your time.

**Comment by malcook** [comment]
I'm trying to follow this issue since I am current challenged in using the bootstrap method by what I thought was an unrelated issue (#1147) but now I wonder two things:

First:

Should I be using --without-tcl if I (currently) intend only to use Lmod - and if so how do I specify this via the bootstrap method, if possible

Or should I want to change my intentions and use tcl version of modules, and install it before easyBuild?

Second:

I am confused by  "I've updated the EB docs to mention that Tcl is also required with Lmod in case you'll be using Tcl module files (only option with EB, for now)"... 

Does the 'only option with EB' mean that EasyBuild only works with Tcl modules, but not Lmod ones?   I hope I am just misunderstanding this sentence.  Am I?

I th

**Comment by malcook** [comment]
@boegel @rtmclay -- 
I think @sylmarien need not have installed any other modules implementation.  

Instead, the lmod implementation should be enough.  

At least, this is one thing I learned and reported in # 1147

In short, what he needs to do before trying to install easyBuild is call

```
module load lmod
```
which which will put lmod/libexec/lmod on $PATH so it can be found when sought during bootstrap

But I still have a follow-on problem with that issue, if anyone has a suggestion for me I'd be happy to be able to use the bootstrap method.

**Comment by boegel** [comment]
@malcook: Seems like this was left unanswered, my apologies.

To clarify: EasyBuild (for now) only *produce* Tcl module files. Lmod consumes module files in both Tcl and Lua syntax. In the future EasyBuild will also be able to produce Lua module files (see https://github.com/hpcugent/easybuild-framework/pull/1060).
So, for now, if you want to combine EasyBuild and Lmod, Lmod should be configured such that it can consume Tcl module files (which is the default). This means `tclsh` must also be available, as you discovered.

EasyBuild requires that the `lmod` binary is available in `$PATH`, this is now (more) clearly mentioned in the documentation, see http://easybuild.readthedocs.org/en/latest/Installation.html#required-modules-tool.

The bootstrap script has recently been made a lot more robust (see for example https://github.com/hpcugent/easybuild-framework/pull/1215).
@sylmarien: care to give it another try, with a fresh download from https://github.com/hpcugent/easybuild-framework/tree/develop/easybuild/scripts/bootstrap_eb.py?

Unless there's anything else unclear, I propose we close this issue. All agreed?

**Comment by malcook** [comment]
+1 to close.  @sylmarien ?

**Comment by sylmarien** [comment]
Sorry for the delay, as far as I'm concerned the answer is satisfying (the documentation states that Lmod also requires Tcl), so I don't mind closing this bug either.

But for the bootstrap matter, I tried it reproducing the same sort of configuration I had when I opened this issue (so I purposely uninstalled Tcl packages) and try to install EasyBuid with the Bootstrap method.

It crashed with the following output: [gist](https://gist.github.com/sylmarien/e2a9ee734ebf57e285b2)

It is the same error that was encountered at the beginning.

I don't know if the bootstrap was supposed to catch the error and tell me to install Tcl or if it is the intended behavior. Even though it would be easier if the error message referred to the Tcl problem, I don't think it is mandatory since the documentation now state that it is a dependency.

On a side note, I didn't update Lmod, so maybe you rely on something that was added/corrected after the creation of this ticket (as I understand its behavior has been changed when checking for tclsh) and that isn't implemented on my version (the Debian package: Lmod 5.8-1).

To clarify, the bootstrap didn't catch the missing dependency, but I think it is not a big deal and that we can close this bug since the documentation is clear on that point.

I close, don't hesitate to re-open if you think there is more to discuss.
